edge-hub/src/Microsoft.Azure.Devices.Routing.Core/endpoints/StoringAsyncEndpointExecutor.cs

```diff
@@ -23,51 +25,50 @@ public class StoringAsyncEndpointExecutor : IEndpointExecutor
         readonly IMessageStore messageStore;
         readonly Task sendMessageTask;
         readonly AsyncManualResetEvent hasMessagesInQueue = new AsyncManualResetEvent(true);
-        readonly ICheckpointer checkpointer;
-        readonly AtomicReference<IList<uint>> priorities = new AtomicReference<IList<uint>>(new List<uint>());
         readonly AsyncEndpointExecutorOptions options;
-        readonly EndpointExecutorFsm machine;
         readonly CancellationTokenSource cts = new CancellationTokenSource();
+        readonly ICheckpointerFactory checkpointerFactory;
+        readonly EndpointExecutorConfig config;
+        AtomicReference<OrderedDictionary> prioritiesToFsms;
+        EndpointExecutorFsm lastUsedFsm;
 
         public StoringAsyncEndpointExecutor(
             Endpoint endpoint,
-            IList<uint> priorities,
-            ICheckpointer checkpointer,
+            ICheckpointerFactory checkpointerFactory,
             EndpointExecutorConfig config,
             AsyncEndpointExecutorOptions options,
             IMessageStore messageStore)
         {
-            Preconditions.CheckNotNull(endpoint);
-            Preconditions.CheckNotNull(config);
-            Preconditions.CheckNotNull(priorities);
-            Preconditions.CheckArgument(priorities.Count != 0);
-            this.checkpointer = Preconditions.CheckNotNull(checkpointer);
+            this.Endpoint = Preconditions.CheckNotNull(endpoint);
+            this.checkpointerFactory = Preconditions.CheckNotNull(checkpointerFactory);
+            this.config = Preconditions.CheckNotNull(config);
             this.options = Preconditions.CheckNotNull(options);
-            this.machine = new EndpointExecutorFsm(endpoint, checkpointer, config);
             this.messageStore = messageStore;
             this.sendMessageTask = Task.Run(this.SendMessagesPump);
-
-            this.UpdatePriorities(priorities);
+            this.prioritiesToFsms = new AtomicReference<OrderedDictionary>(new OrderedDictionary());
         }
 
-        public Endpoint Endpoint => this.machine.Endpoint;
+        public Endpoint Endpoint { get; }
 
-        public EndpointExecutorStatus Status => this.machine.Status;
+        public EndpointExecutorStatus Status => this.lastUsedFsm.Status;
```

Comment:
Can you find out exactly what the status is being used for please?

Before, the status would answer "what is the status of the EndpointExecutor's FiniteStateMachine"

Now, status is answering "What is the status of the last used FiniteStateMachine by the EndpointExecutor"

It's very similar, because it will still be true that the last used FSM's status is probably what we care about, and all the other FSM's are Idle by definition, but to me it seems more awkward now to ask the status like this.

Reply:
I followed the call stack, looks like it's only called by tests now.  Maybe something called it before.  We can remove this if we ever get around to cleaning up the SyncEndpointExecutor and AsyncEndpointExecutor implementations, both of which are not valid for Edge scenarios.

Reply:
Since it's unused, can you add those as a task, and add removing this as a part of those tasks as well please, so we don't forget

Reply:
6470923 - Remove unused message forwarding modes

---
In reply to: [391714814](https://github.com/Azure/iotedge/pull/2670#discussion_r391714814) [](ancestors = 391714814)